File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(table, condition)
    for _, value in ipairs(table) do
        if condition(value) then
            return string.lower(value)
        end
    end
    return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
    return coroutine.create(function()
        local power = 1
        while power < limit do
            coroutine.yield(power)
            power = power * base
        end
    end)
end

-- Write your say function here
function say(word)
    local words = {}    
    local function addToSentence(next_word)
        if next_word == nil then
            return table.concat(words, " ")
        else
            table.insert(words, next_word)
            return addToSentence
        end
    end
    if word == nil then
        return addToSentence()
    else
        table.insert(words, word)
        return addToSentence
    end
end

-- Write your line count function here
function meaningful_line_count(filename)
    local file, err = io.open(filename, "r")
    if not file then
        error("No such file: " .. filename)
    end
    local count = 0
    for line in file:lines() do
      -- effectively trims leading and trailing whitespace from line, leaving just the core content of the string. I used chatGPT to help me figure this part out. 
        line = line:gsub("^%s*(.-)%s*$", "%1")
        if line ~= "" and not line:match("^#") then
            count = count + 1
        end
    end
    file:close()
    return count
end

-- Write your Quaternion table here

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Tuple


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts

# Write your first then lower case function here
def first_then_lower_case(seq, predicate):
    for item in seq:
        if predicate(item):
            return item.lower()
    return None

# Write your powers generator here
def powers_generator(*, base: int, limit: int):
    value = 1
    while value <= limit:
        yield value
        value *= base

# Write your say function here
def say(word = None):
    if 'words' not in say.__dict__:
        say.words = []
    
    if word is None:
        result = ' '.join(say.words)
        say.words = [] 
        return result
    
    say.words.append(word)
    return say

# Write your line count function here
def meaningful_line_count(filename):
    try:
        with open(filename, 'r') as file:
            count = 0
            for line in file:
                line = line.strip()
                if line and not line.startswith('#'):
                    count += 1
        return count
    except FileNotFoundError:
        raise FileNotFoundError(f"No such file: {filename}")

# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion:
    a: float  
    b: float  
    c: float  
    d: float  

    @property
    def coefficients(self) -> Tuple[float, float, float, float]:
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self) -> 'Quaternion':
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, other: 'Quaternion') -> 'Quaternion':
        return Quaternion(
            self.a + other.a,
            self.b + other.b,
            self.c + other.c,
            self.d + other.d
        )

    def __mul__(self, other: 'Quaternion') -> 'Quaternion':
        a1, b1, c1, d1 = self.a, self.b, self.c, self.d
        a2, b2, c2, d2 = other.a, other.b, other.c, other.d

        return Quaternion(
            a1*a2 - b1*b2 - c1*c2 - d1*d2,
            a1*b2 + b1*a2 + c1*d2 - d1*c2,
            a1*c2 - b1*d2 + c1*a2 + d1*b2,
            a1*d2 + b1*c2 - c1*b2 + d1*a2
        )

    def __eq__(self, other: 'Quaternion') -> bool:
        return (self.a, self.b, self.c, self.d) == (other.a, other.b, other.c, other.d)

    def __str__(self) -> str:
        parts = []
        if self.a != 0:
            parts.append(f"{self.a}")
        if self.b != 0:
            if self.b == 1:
                parts.append("i")
            elif self.b == -1:
                parts.append("-i")
            else:
                parts.append(f"{self.b}i")
        if self.c != 0:
            if self.c == 1:
                parts.append("j")
            elif self.c == -1:
                parts.append("-j")
            else:
                parts.append(f"{self.c}j")
        if self.d != 0:
            if self.d == 1:
                parts.append("k")
            elif self.d == -1:
                parts.append("-k")
            else:
                parts.append(f"{self.d}k")

        if not parts:
            return "0"

        return "+".join(parts).replace("+-", "-")